src/Middleware/RequestThrottling/src/Internal/RequestQueue.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.RequestThrottling.Internal
+{
+    internal class RequestQueue : IDisposable
+    {
+        private SemaphoreSlim _semaphore;
+        private object _waitingRequestsLock = new object();
+        public readonly int MaxConcurrentRequests;
+        public int WaitingRequests { get; private set; }
+
+        public RequestQueue(int maxConcurrentRequests)
+        {
+            MaxConcurrentRequests = maxConcurrentRequests;
+            _semaphore = new SemaphoreSlim(maxConcurrentRequests);
+        }
+
+        public async Task EnterQueue()
+        {
+            var waitInQueueTask = _semaphore.WaitAsync();
+
+            var needsToWaitOnQueue = !waitInQueueTask.IsCompletedSuccessfully;
+            if (needsToWaitOnQueue)
+            {
+                lock (_waitingRequestsLock)
+                {
+                    WaitingRequests++;
```

Comment:
Why not `Interlocked.Increment`?

Reply:
Not needed since the variable is only accessed inside a lock

Reply:
And the lock only protects the variable. Interlocked.Increment would be more efficient.

Reply:
Good catch! Interlocked has better performance, and I'll go add it in the next PR